fix: allows deleted users to signup with same email

shreyanshdwivedi opened the pull request:
<!--
(Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.)
-->
<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->

Fixes #5889 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes <!-- use `nosetests tests/all` to run all the tests -->
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
<!-- If an existing function does not have a docstring, please add one -->
- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
Currently, if a user deletes his/her account or the account is deleted by an admin, he/she can't signup with same email.

#### Changes proposed in this pull request:
 - Revamps the deletion of a user on server.
 - adds a function to add `.deleted` to a user's email which is to be deleted
 - removes `.deleted` from user's email which is to be restored.

shreyanshdwivedi commented:
@uds5501 @mrsaicharan1 @iamareebjamal  everything is working fine but when I update the email of user in `modify_email_for_user_to_be_restored` function, it is not getting reflected on database. Can you please help me out with this?

iamareebjamal commented on app/api/helpers/user.py:
```diff
@@ -0,0 +1,49 @@
+            if userEmail.endswith('.deleted'):
+                userEmail = userEmail[:-8]
+            else:
+                userEmail = userEmail + '.new'
```
Why?

shreyanshdwivedi replied:
Suppose a user is restored but he/she already signed up so the constraint is violated as 2 users cannot have same email. So I just added a '.new' to the email just for this sake.
This case however may happen only once or twice

iamareebjamal replied:
This should not be automated. Admin should first change the email manually to something and then restore

shreyanshdwivedi replied:
@iamareebjamal I'll remove this condition and if you want I can open a separate issue to handle this. 
>when I update the email of user in modify_email_for_user_to_be_restored function, it is not getting reflected on database

In the meantime can you please give this a look? ^^

shreyanshdwivedi replied:
@iamareebjamal any help?

iamareebjamal replied:
Will look tomorrow. @prateekj117 @uds5501 Can you please take a look as well?

uds5501 replied:
@iamareebjamal @shreyanshdwivedi will take a look tonight

uds5501 replied:
@shreyanshdwivedi The code looks correct. Any error logs during implementation?

shreyanshdwivedi replied:
No error logs. @uds5501
@iamareebjamal can you take a look as well please

shreyanshdwivedi commented:
@iamareebjamal @uds5501  I've fixed the issue due to which the email was not getting restored while restoring the user. Also, I've added a migration file to update the email of past deleted users. 
Please review

shreyanshdwivedi commented:
@iamareebjamal @CosmicCoder96 please review this :)

iamareebjamal commented on app/api/helpers/user.py:
```diff
@@ -0,0 +1,47 @@
+            not_unique_email = False
+        else:
+            if user_email.endswith('.deleted'):
+                user_email = user_email[:-8]
```
Why do we need this?

shreyanshdwivedi replied:
When the user is deleted, `.deleted` is added so when the deleted user is restored, the original email should be restored which is substring obtained by omitting last 8 characters i.e. `.deleted` @iamareebjamal

iamareebjamal replied:
Not maintainable. Remove the loop. Remove '.deleted' from the last. If you have to use :-8, then extract '.deleted' in a variable and use len(...)

If there's an error restoring the user in case after '.deleted' is removed, it is still not unique, throw an error and tell the admin that there is already an email registered on this name. Manually edit that and try restoring

You should not automate human decisions

shreyanshdwivedi replied:
@iamareebjamal I've updated the PR. Please take a look at the updated conditions now

shreyanshdwivedi commented:
@iamareebjamal please take a look at the updated conditions and review the PR :)

iamareebjamal commented on app/api/helpers/user.py:
```diff
@@ -0,0 +1,47 @@
+        else:
+            return False, user
+        return True, user
+    return False, user
```
Don't return tuple, raise an exception if problem is encountered

shreyanshdwivedi commented:
@iamareebjamal please review. I've updated the conditions

iamareebjamal commented on app/api/helpers/user.py:
```diff
@@ -0,0 +1,49 @@
+        else:
+            raise ForbiddenException({'pointer': '/data/attributes/email'},
+                                     "This email is already registered! Manually edit and then try restoring")
+    raise ForbiddenException({'pointer': '/data/attributes/email'},
```
This will be raised unconditionally

shreyanshdwivedi commented:
@iamareebjamal removed the duplicate error message. Please review

iamareebjamal commented on app/api/helpers/user.py:
```diff
@@ -0,0 +1,48 @@
+    """
+    user_email = user.email
+    if user_email.endswith('.deleted'):
+        remove_str = '.deleted'
```
Move this statement a line above and replace '.deleted' with remove_str in if statement. That was the point of extracting it to a variable

shreyanshdwivedi commented:
@iamareebjamal updated the PR. Please review

iamareebjamal commented:
Build is failing

shreyanshdwivedi commented:
@iamareebjamal this is some random failure which occurs on other PRs too :sweat_smile: 
```
error: Hooks handler process 'dredd-hooks-python ./tests/hook_main.py' exited with status: 1
warn: Hook handling timed out.
error: Hooks handler process 'dredd-hooks-python ./tests/hook_main.py' exited with status: 1
info: Backend server process exited
The command "dredd" failed and exited with 1 during .
```

iamareebjamal commented:
Multiple Migration Heads. Not a random error

uds5501 commented:
@shreyanshdwivedi It failed again. OH. kindly change the `revises` of your migration file to the current head. :)

shreyanshdwivedi commented:
@uds5501 yep. Was fixing that only :sweat_smile: 
Good to know the real cause of such errors. Thanks @iamareebjamal 
Let's wait for the test to pass

iamareebjamal commented:
https://github.com/fossasia/open-event-server/blob/development/scripts/test_multiple_heads.sh is not working correctly. Open an issue

shreyanshdwivedi commented:
@iamareebjamal tests passed. Please merge
